doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

```diff
@@ -0,0 +1,92 @@
+
+Here's an example of the upsert behavior where the source consists of a key, value and the timestamp from metadata and is ordered by the timestamp.
+
+```
+ Previously persisted data        Incoming kafka stream decoded
++------------------------+         +-------------------------+
+| Ok (key1, value1, 100) |         |  Ok (key1, value3, 200) |
+| Ok (key2, value2, 201) |         |  Err(key3, some_error2) |
+| Err(key3, some_error1) |         |  Ok (key2, value4, 200) |
++------------------+-----+         +-----+-------------------+
+                   |                     |
+                   |                     |
+                   |                     |
+                   |                     |
+                   v    Upsert Output    v
+                  +------------------------+
+                  | Ok (key1, value3, 200) |
+                  | Ok (key2, value2, 201) |
+                  | Err(key3, some_error2) |
+                  +------------------------+
+```
+Note: As shown in the example above, the errors are still implicitly ordered by offset as we do not persist any extra metadata for them separately. A later error with the same key will always overwrite a previous one.
+
+# Rollout
+This feature will be rolled out behind the unsafe flag.
+
+## Testing and Observability
+
+The custom ordering can be tested via testdrive tests in the CI.
+
+## Lifecycle
+
+What will it take to promote this to the next stage i.e. Alpha?
+
+- We should work with some users and see how useful this is to them. This would also give us a better idea on how to expand this feature later on.
+- Some questions to give us a better idea
+    - How is their data structured and what columns will they want to use for the custom order by?
+    - What kind of compaction are they using?
+
+# Drawbacks
+
+Users will need to be careful of what kind of compaction they have for the kafka topic. The upsert will do the correct thing based on the data it sees.
+
+If old data has been compacted away either via time based or key based compaction, and we start ingesting the source after that, the result of upsert might not be accurate.
+
+If users actually are using this custom order by, it’s likely they are **not** using key based compaction with Kafka because semantically that would be upsert-ing on offset.
+
+# Conclusion and alternatives
+
+An alternative could be waiting for `TRANSFORM USING` which has a much broader scope and could cover this particular scenario. In the meantime, having a smaller scoped custom order by could be useful for interested users.
+
+# Unresolved questions
+- In case the order by value of two entries are same, should it chose one of them deterministically? For eg, say we already have (key1, value1, ts=100) and have incoming (key1, value2, ts=100). Should we override `value1` with `value2` or not?
+
+# Future work
+
+Depends upon user feedback to see how this should be expanded.
```

Comment:
I think you can just mention that adding support for user-provided columns is a future extension, but dealing with tombstones is an unresolved problem!

Reply:
Updated!